WordPress/Classes/Utility/DiskCache.swift

```diff
@@ -0,0 +1,273 @@
+        guard let url = fileURL(for: key) else { return }
+        try? FileManager.default.removeItem(at: url)
+    }
+
+    /// Removes all cached entries.
+    public func removeAll() {
+        do {
+            try FileManager.default.removeItem(at: rootURL)
+            try FileManager.default.createDirectory(at: rootURL, withIntermediateDirectories: true, attributes: nil)
+        } catch {
+            DDLogError("Failed to clear cache with error: \(error)")
+        }
+    }
+
+    /// Returns the URL for the given cache key.
+    public nonisolated func fileURL(for key: String) -> URL? {
+        guard let filename = key.sha1 else { return nil }
+        return rootURL.appendingPathComponent(filename, isDirectory: false)
+    }
+
+    // MARK: Sweep
+
+    private func performAndScheduleNextSweep() {
+        sweep()
+        DispatchQueue.main.asyncAfter(deadline: .now() + configuration.sweepInterval) { [weak self] in
+            Task {
+                await self?.performAndScheduleNextSweep()
+            }
+        }
+    }
+
+    /// Performs cache sweep and removes the least recently items which no longer fit.
+    public func sweep() {
+        var items = contents(keys: [.contentAccessDateKey, .totalFileAllocatedSizeKey])
+        guard !items.isEmpty else {
+            return
+        }
+        var size = items.reduce(0) { $0 + ($1.meta.totalFileAllocatedSize ?? 0) }
+
+        guard size > configuration.sizeLimit else {
+            return // All good, no need to perform any work.
+        }
+
+        let targetSizeLimit = Int(Double(configuration.sizeLimit) * configuration.trimRatio)
+
+        // Most recently accessed items first
+        let past = Date.distantPast
+        items.sort { // Sort in place
+            ($0.meta.contentAccessDate ?? past) > ($1.meta.contentAccessDate ?? past)
+        }
+
+        // Remove the items until it satisfies both size and count limits.
+        while size > targetSizeLimit, let item = items.popLast() {
+            size -= (item.meta.totalFileAllocatedSize ?? 0)
+            try? FileManager.default.removeItem(at: item.url)
+        }
+    }
+
+    // MARK: Contents
+
+    struct Entry {
+        let url: URL
+        let meta: URLResourceValues
+    }
+
+    func contents(keys: [URLResourceKey] = []) -> [Entry] {
+        guard let urls = try? FileManager.default.contentsOfDirectory(at: rootURL, includingPropertiesForKeys: keys, options: .skipsHiddenFiles) else {
+            return []
+        }
+        let keys = Set(keys)
+        return urls.compactMap {
+            guard let meta = try? $0.resourceValues(forKeys: keys) else {
+                return nil
+            }
+            return Entry(url: $0, meta: meta)
+        }
+    }
+
+    // MARK: Inspection
+
+    /// The total number of items in the cache.
+    ///
+    /// - important: Requires disk IO, avoid using from the main thread.
+    public var totalCount: Int {
```

Comment:
If we don't want this used from the main thread, we should enforce it in code.

Additionally, it's generally preferred to avoid using `var` access for expensive operations – this might make more sense as a method like `countItems()`? WDYT?

Reply:
It can't be used from the main thread because it's actor isolated, but I see your point about the naming. I agree that It should probably be a method.

Reply:
Updated